torchao/quantization/GPTQ.py

```diff
@@ -0,0 +1,579 @@
+
+from typing import Optional
+
+import torch
+
+import torch.fx as fx
+import torch.nn as nn
+import torch.nn.functional as F
+# from model import Transformer  # pyre-ignore[21]
+from torch.utils._pytree import tree_flatten, tree_unflatten
+
+aten = torch.ops.aten
+
+## generate.py ##
+
+
+def encode_tokens(tokenizer, string, bos=True, device="cuda"):
+
+    tokens = tokenizer.encode(string)
+    if bos:
+        tokens = [tokenizer.bos_id()] + tokens
+    return torch.tensor(tokens, dtype=torch.int, device=device)
+
+
+def model_forward(model, x, input_pos):
+    return model(x, input_pos)
+
+
+## eval.py ##
+
+try:
+    import lm_eval  # pyre-ignore[21]  # noqa: F401
+
+    lm_eval_available = True
+except:
+    lm_eval_available = False
+
+if lm_eval_available:
+    try:  # lm_eval version 0.4
+        from lm_eval.evaluator import evaluate  # pyre-ignore[21]
+        from lm_eval.models.huggingface import HFLM as eval_wrapper  # pyre-ignore[21]
+        from lm_eval.tasks import get_task_dict  # pyre-ignore[21]
+    except:  # lm_eval version 0.3
+        from lm_eval import base, evaluator, tasks
+
+        eval_wrapper = base.BaseLM
+        get_task_dict = tasks.get_task_dict
+        evaluate = evaluator.evaluate
+
+
+def setup_cache_padded_seq_input_pos_max_seq_length_for_prefill(
+    model: torch.nn.Module,  # pyre-ignore[11]
+    prompt: torch.Tensor,
+    max_new_tokens: int,
+    max_seq_length: Optional[int] = None,
+    block_size: int = 2048,
+):
+    """
+    Sets up model cache and does some bookkeeping calculations for prompt, input_pos and max_seq_length
+    that are needed for prefill or model_forward
+
+    Args:
+        model (torch.nn.Module): The model whose cache gets set up
+        prompt (torch.Tensor): Tensor of shape (T) with indices of the prompt sequence.
+        max_new_tokens (int): The desired maximum number of new tokens that can be generated.
+        max_seq_length (Optional[int], optional): The maximum sequence length allowed.
+
+    Returns:
+        seq (torch.Tensor): prompt but padded with zeros to size max_seq_length
+        input_pos (torch.Tensor): tensor of integers in increasing order
+        max_seq_length (int): The maximum sequence length allowed, updated based on other numbers
+    """
+    T = prompt.size(0)
+    T_new = T + max_new_tokens
+    if max_seq_length is None:
+        max_seq_length = min(T_new, block_size)
+
+    device, dtype = prompt.device, prompt.dtype
+    # create an empty tensor of the expected final shape and fill in the current tokens
+    empty = torch.empty(T_new, dtype=dtype, device=device)
+    empty[:T] = prompt
+    seq = empty
+    input_pos = torch.arange(0, T, device=device)
+
+    # no caches in executorch llama2 7b model?
+    # with torch.device(device):
+    #     model.setup_caches(max_batch_size=1, max_seq_length=max_seq_length)
+
+    return seq, input_pos, max_seq_length
+
+
+class GPTFastEvalWrapper(eval_wrapper):  # pyre-ignore[11]
```

Comment:
So this is essentially just the quantizer from gpt-fast?

Reply:
these are just copy pasting btw. this is a eval wrapper, used to record inputs for GPTQ I think